mpqp/qasm/open_qasm_2_and_3.py

```diff
@@ -740,3 +760,333 @@ def remove_include(qasm_code: str) -> str:
+    elif instr_name == "u3":
+        header_code += add_qe_lib()
+        instructions_code += re.sub(r"\bu3\b", "u", instr) + ";\n"
+    elif instr_name == "cp":
+        header_code += add_qe_lib()
+        instructions_code += re.sub(r"\bcp\b", "cu1", instr) + ";\n"
+    elif instr_name in std_gates_3_to_2_map:
+        converted_instr_name = std_gates_3_to_2_map[instr_name]
+        instructions_code += (
+            re.sub(r"\b" + instr_name + r"\b", converted_instr_name, instr) + ";\n"
+        )
+    elif instr_name in std_gates_3:
+        m = re.match(r"\s*(.*)", instr)
+        if m:
+            instructions_code += m.group(1) + ";\n"
+    elif instr_name == "gate":
+        m = re.match(
+            r"\s*gate\s+(\w+)\s*(\(([^)]*)\))?\s*([\w\s,]*)\s*{\s*([^}]*)\s*}",
+            instr,
+            re.DOTALL,
+        )
+        if m:
+            gate_name = m.group(1)
+            params = f"({m.group(3)})" if m.group(3) else ""
+            qubits = m.group(4).strip()
+            body = m.group(5)
+
+            defined_gates.add(gate_name)
+            g_string = f"gate {gate_name}{params} {qubits} {{\n"
+            g_instructions = filter(
+                lambda i: not re.fullmatch(r"\s*", i), body.split(";")
+            )
+            for instruction in g_instructions:
+                instruction = instruction.strip()
+                i_code, h_code = convert_instruction_3_to_2(
+                    instruction,
+                    included_instr,
+                    included_tree_current,
+                    defined_gates,
+                    path_to_main,
+                )
+                g_string += " " * 4 + i_code  # Add indentation to body instructions
+                header_code += h_code
+
+            # Finalize the gate and add it to instructions_code
+            instructions_code += g_string + "}\n"
+
+    elif re.match(r"\s*if\s*\(.*?\)\s*.+", instr):
+        m = re.match(r"(\s*if\s*\(.*?\))\s*(.+)", instr)
+        if m:
+            if_statement = m.group(1)
+            nested_instr = m.group(2)
+            i_code, h_code = convert_instruction_3_to_2(
+                nested_instr,
+                included_instr,
+                included_tree_current,
+                defined_gates,
+                path_to_main,
+            )
+            instructions_code += if_statement + " " + i_code
+            header_code += h_code
+    elif instr_name in {"reset", "barrier"}:
+        instructions_code += instr + ";\n"
+    else:
+        gate = instr.split()[0].split("(")[0]
+        if gate == "ctrl":
+            gate = instr.split()[2]
+        if gate not in defined_gates:
+            raise ValueError(
+                f"Gates undefined at the time of usage: {gate}, {instr_name}"
+            )
+        m = re.match(r"\s*(.*)", instr)
+        if m:
+            instructions_code += m.group(1) + ";\n"
+
+    return instructions_code, header_code
+
+
+@typechecked
+def open_qasm_3_to_2(
```

Comment:
When we translate a circuit from OpenQASM3 to OpenQASM2, we lose the information of the global phase of the circuit. I think we should be able to estimate it through the conversions, and return it with the string (so it can be stored in a circuit, or post-processed by the user if needed).

Reply:
done